pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java

```diff
@@ -0,0 +1,844 @@
+            int roleId = studyRoles.get(role).roleId;
+            try {
+                addUserRole(userId, roleId, groupId);
+                String msg = String.format("Added role %s for user %s in study group %s", role, email, studyGroup);
+                log.info(msg);
+            } catch (Exception e) {
+                String msg = String.format("Error adding user %s to role %s for study group %s", email, role, studyGroup);
+                throw new DsmInternalError(msg, e);
+            }
+        }
+    }
+
+    /**
+     * Remove roles for list of users
+     */
+    public void removeUserRoles(UserRoleRequest req) {
+        validateRoleRequest(req);
+
+        int groupId = verifyStudyGroup(studyGroup);
+        Map<String, RoleInfo> studyRoles = getAdminRoles(groupId);
+        List<String> roleNames = req.getRoles();
+        validateRoles(roleNames, studyRoles.keySet());
+
+        for (String userEmail: req.getUsers()) {
+            if (StringUtils.isBlank(userEmail)) {
+                throw new DSMBadRequestException("Invalid user email: blank");
+            }
+            removeUserRoles(userEmail, roleNames, groupId, studyRoles);
+        }
+    }
+
+    protected void removeUserRoles(String email, List<String> roles, int groupId, Map<String, RoleInfo> studyRoles) {
+        // TODO loop is similar to addUserRoles -DC
+        int userId = verifyUserByEmail(email, groupId);
+
+        for (String role : roles) {
+            if (StringUtils.isBlank(role)) {
+                throw new DsmInternalError("Invalid role: blank");
+            }
+            // role names already validated
+            int roleId = studyRoles.get(role).roleId;
+            try {
+                // ignore failure to remove
+                deleteUserRole(userId, roleId, groupId);
+                String msg = String.format("Removed role %s for user %s in study group %s", role, email, studyGroup);
+                log.info(msg);
+            } catch (Exception e) {
+                String msg = String.format("Error removing user %s from role %s for study group %s", email, role, studyGroup);
+                throw new DsmInternalError(msg, e);
+            }
+        }
+    }
+
+    protected void validateRoleRequest(UserRoleRequest req) {
+        if (CollectionUtils.isEmpty(req.getUsers())) {
+            throw new DSMBadRequestException("Invalid users: empty");
+        }
+        if (CollectionUtils.isEmpty(req.getRoles())) {
+            throw new DSMBadRequestException("Invalid roles: empty");
+        }
+    }
+
+    protected void validateRoles(List<String> roleNames, Set<String> validRoleNames) {
+        if (validRoleNames.containsAll(roleNames)) {
+            return;
+        }
+        Collection<String> badRoles = CollectionUtils.subtract(roleNames, validRoleNames);
+        String msg = String.format("Invalid roles for study group %s: %s", studyGroup, String.join(",", badRoles));
+        throw new DSMBadRequestException(msg);
```

Comment:
Am I correct that this is checking that all the roles do belong to this study group?

Reply:
'validRoleNames' are all the roles the operator/admin can manage. If the operator is `pepper_admin` then that is all the study roles. If the operator is `study_user_admin` then it is less than all the study roles -- for instance, no GP only roles, no `study_user_admin`, etc.